### ISSUE TITLE
`optimisticData(old)` in `useSWRMutation` is not invoked with latest data from cache 

### ISSUE THREAD

**Issue by westmark** [vercel/swr#2701]
# Bug report

## Description / Observed Behavior

When calling the trigger function returned from `useSWRMutation` in rapid succession(i.e. before the fetcher has returned) and using the `optimisticData` callback, the callback does not receive the latest data from the cache, or perhaps more importantly: it does not receive the data that was returned from its last invokation if there have been such and the fetcher is still unresolved. 

## Expected Behavior

I expect the data sent to `optimisticData` to always be the latest (optimistic) data

## Repro Steps / Code Example

https://codesandbox.io/s/aged-shadow-l4wrhj?file=/src/App.js

Notice how the counter changes immediately to 2 but does not progress to 3 until the clicking stops and the fetcher is allowed to return

## Additional Context

SWR version 2.1.1 (also in 2.2 in my project)

Not sure if you consider this a bug or a feature, considering the following text in the docs

> Also, this hook doesn’t share states with other useSWRMutation hooks.

However, does this count as an "other hook"?

**Comment by henriqemalheiros** [comment]
Possibly related to https://github.com/vercel/swr/issues/2186? I believe it's the same reasoning and both issues require proper enqueuing of `mutate` calls.

**Comment by JamieS1211** [comment]
Just worked around this myself by using a "boundMutate" and using the data form "useSWR" hook so I the data doesn't come from the function call on updating optimistic data.

**Comment by PieterDePauw** [comment]
> Just worked around this myself by using a "boundMutate" and using the data form "useSWR" hook so I the data doesn't come from the function call on updating optimistic data.

Can you provide an example of how you solved this problem ?

**Comment by alexnsorensen** [comment]
Is there an update on getting this fixed as it's a bit of a deal breaker for me using this in my project.

**Comment by alexnsorensen** [comment]
> boundMutate

I tried this, and it's a good work around, thank you for that.

**Comment by alexnsorensen** [comment]
> Just worked around this myself by using a "boundMutate" and using the data form "useSWR" hook so I the data doesn't come from the function call on updating optimistic data.

@JamieS1211 I tried this, and it's a good work around for what I'm after, thanks.

**Comment by JamieS1211** [comment]
> > Just worked around this myself by using a "boundMutate" and using the data form "useSWR" hook so I the data doesn't come from the function call on updating optimistic data.
> 
> @JamieS1211 Can you provide an example of how you solved this problem ?

Sorry missed this - the documentation on bound mutation can be found here [https://swr.vercel.app/docs/mutation.en-US#bound-mutate](https://swr.vercel.app/docs/mutation.en-US#bound-mutate). Essentially use that, then when an optimistic update happens, the `data` variable from the hook updates and causes a re-render. Then you can do a mutation with the optimistic data present. In the example they show the mutation as: `mutate({ ...data, name: newName })` however you can do as needed here.

**Comment by PieterDePauw** [comment]
> > > Just worked around this myself by using a "boundMutate" and using the data form "useSWR" hook so I the data doesn't come from the function call on updating optimistic data.
> > 
> > 
> > @JamieS1211 Can you provide an example of how you solved this problem ?
> 
> Sorry missed this - the documentation on bound mutation can be found here [swr.vercel.app/docs/mutation.en-US#bound-mutate](https://swr.vercel.app/docs/mutation.en-US#bound-mutate). Essentially use that, then when an optimistic update happens, the `data` variable from the hook updates and causes a re-render. Then you can do a mutation with the optimistic data present. In the example they show the mutation as: `mutate({ ...data, name: newName })` however you can do as needed here.

So, if I am not mistaken, you essentially trigger a revalidation in between consecutive mutations?

**Comment by alexnsorensen** [comment]
> > > > Just worked around this myself by using a "boundMutate" and using the data form "useSWR" hook so I the data doesn't come from the function call on updating optimistic data.
> > > 
> > > 
> > > @JamieS1211 Can you provide an example of how you solved this problem ?
> > 
> > 
> > Sorry missed this - the documentation on bound mutation can be found here [swr.vercel.app/docs/mutation.en-US#bound-mutate](https://swr.vercel.app/docs/mutation.en-US#bound-mutate). Essentially use that, then when an optimistic update happens, the `data` variable from the hook updates and causes a re-render. Then you can do a mutation with the optimistic data present. In the example they show the mutation as: `mutate({ ...data, name: newName })` however you can do as needed here.
> 
> So, if I am not mistaken, you essentially trigger a revalidation in between consecutive mutations?

If I'm not mistaken this is what @JamieS1211 is doing as this is what worked for me:
`// In my example below, data is an array
const { data, mutate } = useSWR(['keyHere']);

function saveData(newData) { 
	mutate(/* Post here to server */,
	{
		populateCache: false,
		optimisticData: () => {
			return [...data, newData];
		}
	);
}`

**Comment by JamieS1211** [comment]
Yep thats it, not sure if there is some drawback of this approach so maybe a maintainer could chip in on that front

**Comment by hitsthings** [comment]
For anyone else seeing old data passed to your optimisticData function, I had a slightly different issue. I was using a filter function for the keys (e.g. `key => key.startsWith('blah')`). When I instead did my filtering before starting the process, and called mutate for each key individually that seemed to fix it. So old was:

```js
const { mutate } = useSWRConfig();

useEffect(() => {
    setTimeout(onCompletelySeparateEventLoop, 1000);
    
    function onCompletelySeparateEventLoop() {
        const p = doAThing();
        mutate(key => key.startsWith('blah'), doAThing, {
              populateCache: false,
              revalidate: false,
              optimisticData: (cached) => transform(cached)
        }
    } 
}, [])
```

New:

```js
const { mutate, cache } = useSWRConfig();

useEffect(() => {
    setTimeout(onCompletelySeparateEventLoop, 1000);
    const keys = Array.from(cache.keys()).filter(key => key.startsWith('blah'));
    function onCompletelySeparateEventLoop() {
        const p = doAThing();
        for (const key of keys) {
            mutate(key, doAThing, {
                  populateCache: false,
                  revalidate: false,
                  optimisticData: (cached) => transform(cached)
            }
        }
    } 
}, [])
```

This _still_ doesn't get the latest data if you have multiple optimisticData calls in a row - but `cache.get(key)` will have the latest in that case.

**Comment by yellowtailfan** [comment]
It would be great to have an example of how to achieve this use case:
1. Use a bound `useSWR` hook to fetch all the user's items in a single API call
2. User modifies multiple items in a single operation (for example, selecting 10 items and adding a tag)
3. Display is instantly updated for all affected items
4. One API PUT call is made for each affected item to store the change in the database
5. The display is revalidated after all of the PUT calls are complete

Our app has been displaying unstable item states when updating multiple items. I used the work around in https://github.com/vercel/swr/issues/2701#issuecomment-1769328493 and it improves behaviour for multiple item updates in my app. However, it still takes a long time for all of the updates in a set to be displayed.

The work around in @alexnsorensen's comment helped me to achieve all the points except (3). I have a wrapper around the bound `mutate()` and I found it was important that the wrapper was called **without** `await`. Otherwise the whole process is slower and I get multiple revalidations as SWR it less likely to batch them.

**Comment by yellowtailfan** [comment]
I found a recipe that works much better for my use case in https://github.com/vercel/swr/issues/2701#issuecomment-2763115225.

The main change is to wrap our per-item API PUT calls with a fetch wrapper that accepts changes for multiple items and sends one PUT request for each item. The fetch wrapper is given to the bound `mutate` call so that we only have to call `mutate` once when multiple items are changed.

The key point seems to be that SWR expects the fetcher in the `mutate` hook to update the same set of data that data that the fetcher in the `useSWR` retrieves. So the fetch and update should operate symmetrically at the same level -- either both should operate on an item, or both should operate on a list of items. In our case, we had this set up asymmetrically -- `useSWR` was fetching a list of items, but `mutate` was being called once per item update.

Here is an outline of the improved recipe that is working well for me:

1. Use a bound `useSWR` hook to fetch all the user's items in a single API GET call
2. User modifies multiple items in a single operation (for example, selecting 10 items and adding a tag)
3. The multiple item changes are passed to a single bound `mutate` call
4. The `mutate` call is given a fetch wrapper that internally calls the API PUT method on each changed item in parallel
5. The `mutate` call is given an `optimisticData` callback which accepts multiple item changes
6. The `mutate` call also has the `populateCache` option set to `false` as our PUT API doesn't return the updated data (something we could improve in the future)
7. The display updates immediately after the user clicks on a button that changes multiple items 🎉
8. Multiple PUT requests are dispatched
9. After the PUT requests complete, a single GET request is sent to revalidate the user's items

**Comment by EthicalJobsJesse** [comment]
It looks like it may just be a case of documentation needing to be updated. The optimisticData function has two parameters (https://github.com/vercel/swr/blob/ba10cf9ed8ac23f83842dd7f8bc62f110ffac290/src/_internal/types.ts#L362), the first with the `currentData` which I believe is whatever is set in the cache, and a second with `displayedData` which is the cached data plus any optimistic updates that have been applied.

If you switch to using the second param then the quick clicking around should work as your intending.

It feels like it really should have displayedData as the first param, as I'm not sure there are too many cases where you want to do an optimistic update without taking into account previous optimistic updates, but it may have just been to ensure no breaking changes.

**Comment by yellowtailfan** [comment]
@EthicalJobsJesse  That sounds like it would be very useful and it would be great to have it documented.
